onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc

```diff
@@ -1,139 +1,277 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#include "qdq_propagation.h"
+#include "core/optimizer/qdq_transformer/qdq_propagation.h"
 
-#include <deque>
+#include <optional>
 
 #include "core/graph/graph_utils.h"
 #include "core/optimizer/initializer.h"
 #include "core/optimizer/qdq_transformer/qdq_util.h"
 #include "core/optimizer/utils.h"
 
 namespace onnxruntime {
+namespace {
+// like graph_utils::GraphEdge, but also allowing the source end to be an input and the destination end to be an output
+struct ExtendedGraphEdge {
+  enum class End { Source,
+                   Destination };
+
+  struct NodeInfo {
+    NodeIndex node_idx;
+    int arg_idx;
+  };
+
+  std::optional<NodeInfo> src;  // if empty, graph input
+  std::optional<NodeInfo> dst;  // if empty, graph output
+  std::string arg_name;
+
+  bool HasGraphInputOrInitializer() const noexcept { return !src.has_value(); }
+  bool HasGraphOutput() const noexcept { return !dst.has_value(); }
+
+  const std::optional<NodeInfo>& GetNodeInfoAtEnd(End end) const {
+    return end == End::Source ? src : dst;
+  }
+
+  Node* GetNodeAtEnd(Graph& graph, End end) const {
+    if (const auto& node_info = GetNodeInfoAtEnd(end); node_info.has_value()) {
```

Comment:
Should this method be non-const? If not do we need two different GetNodeAtEnd methods?

Reply:
renamed it to GetMutableNodeAtEnd. I think it's ok as a const method as it doesn't actually expose modifiable state of the ExtendedGraphEdge, but rather the non-const Graph parameter